lib/xapi-stdext-encodings/encodings.ml

```diff
@@ -20,142 +20,84 @@ exception UTF8_continuation_byte_invalid
 exception UTF8_encoding_not_canonical
 exception String_incomplete
 
-(* === Utility Functions === *)
-
-let ( +++ ) = Int32.add
-let ( --- ) = Int32.sub
-let ( &&& ) = Int32.logand
-let ( ||| ) = Int32.logor
-let ( <<< ) = Int32.shift_left
-let ( >>> ) = Int32.shift_right_logical
-
 (* === Unicode Functions === *)
 
 module UCS = struct
 
-  let min_value = 0x000000l
-  let max_value = 0x1fffffl
-
   let is_non_character value = false
-                               || (0xfdd0l <= value && value <= 0xfdefl) (* case 1 *)
-                               || (Int32.logand 0xfffel value = 0xfffel) (* case 2 *)
-
-  let is_out_of_range value =
-    value < min_value || value > max_value
-
-  let is_surrogate value =
-    (0xd800l <= value && value <= 0xdfffl)
+                               || (0xfdd0 <= value && value <= 0xfdef) (* case 1 *)
+                               || (Int.logand 0xfffe value = 0xfffe) (* case 2 *)
+                               [@@inline]
 
 end
 
 module XML = struct
 
-  let is_forbidden_control_character value = value < 0x20l
-                                             && value <> 0x09l
-                                             && value <> 0x0al
-                                             && value <> 0x0dl
+  let is_forbidden_control_character value = let value = Uchar.to_int value in
```

Comment:
Just from a grammar perspective, I would use "illegal" instead of "forbidden". We expect most characters to be legal - so we want this predicate to evaluate to false as fast as possible. A character is illegal, if it is a control character that is not tab, nl, or cr. I expect `0x0a` to be the most frequent, so would place this after `value < 0x20`.

To make this predicate as fast as possible, we should order conditions such that most likely

Reply:
I've added a separate fastpath for characters that are >= 0x20 and less <0x80. Those do not need to go through UTF8 decoding (they are ASCII), and would always pas this test as well.

I've renamed 'forbidden' to 'illegal'